### PR TITLE
Fix LogRecord "extra" data leaking between handlers

### PR DESCRIPTION
**Pull request by cosminardeleanu**
Context:
- Let's say i have 3 handlers attached to monolog logger: file handler, slack handler, elastic search handler.
  - Note that each handler can push its own processor, which can alter the LogRecord entity.
- On the elastic search handler, we need extra info, so push a processor to add them,  (stupid example: `$logRecord->extra['new_entry_just_for_elastic_search'] = 'stuff';`
  - This "extra" data i also available for file handler and slack handler.
  - Basically i will get in slack handler and file handler, this extra that was supposed to be only for elastic search handler.

Why this is happening?
- LogRecord is passed by reference, to each handler.
- LogRecord accepts to alter data from `extra`, so if one handlers manages to alter this object, rest of remaining handlers, will use the altered object.

### CODE REVIEW

**Comment by cosminardeleanu** [comment]
@Seldaek ^^

**Comment by aistis-** [comment]
I confirm the issue persists with `3.4.0`. Waiting for the merge 🤞

**Comment by stof** [comment]
@Seldaek with the migration to an object in Monolog 3, this is indeed an issue for handler-level processors rather than logger-level processors. Should the logger clone the record before calling the handler ?

**Comment by Seldaek** [comment]
Yeah I think it'd be better to clone.. both in Logger and in GroupHandler and related handlers which fan out records to nested handlers. Otherwise the problem is just shifted one problem lower.

**Comment by stof** [comment]
indeed, fan-out handlers also need to clone.

**Comment by Seldaek** [comment]
Alright I think with my commit it should take care of it for all cases. I hope I didn't forget any fan-out handler.
